pricegraph/src/orderbook.rs

```diff
@@ -606,6 +670,51 @@ mod tests {
         assert!(!orderbook.is_overlapping());
     }
 
+    #[test]
+    fn detects_overlapping_transitive_orders() {
+        // 0 --1.0--> 1 --0.5--> 2 --1.0--> 3 --1.0--> 4
+        //            ^---------1.0--------/^---0.5---/
+        let mut orderbook = orderbook! {
+            users {
+                @1 {
+                    token 1 => 1_000_000,
+                }
+                @2 {
+                    token 2 => 2_000_000,
+                }
+                @3 {
+                    token 3 => 1_000_000,
+                }
+
+                @4 {
+                    token 4 => 1_000_000,
+                }
+                @5 {
+                    token 3 => 2_000_000,
+                }
+            }
+            orders {
+                owner @1 buying 0 [1_000_000] selling 1 [1_000_000],
+                owner @1 buying 3 [1_000_000] selling 1 [1_000_000],
+                owner @2 buying 1 [1_000_000] selling 2 [2_000_000],
+                owner @3 buying 2 [1_000_000] selling 3 [1_000_000] (500_000),
+
+                owner @4 buying 3 [1_000_000] selling 4 [1_000_000],
+                owner @5 buying 4 [1_000_000] selling 3 [2_000_000],
+            }
+        };
+
+        let overlap = orderbook.reduce_overlapping_transitive_orderbook(1, 2);
+
+        // Transitive order `2 -> 3 -> 1`
+        assert_approx_eq!(overlap.asks[0].buy, 500_000.0);
+        assert_approx_eq!(overlap.asks[0].sell, 500_000.0 / FEE_FACTOR);
+
+        // Transitive order `1 -> 2`
+        assert_approx_eq!(overlap.bids[0].buy, 1_000_000.0);
+        assert_approx_eq!(overlap.bids[0].sell, 2_000_000.0);
```

Comment:
@fleupold Regarding your question about the fees, the example should *hopefully* help clarify:
- The transitive order of `1 -> 2` is just equal to the order, so buying `1_000_000` and selling `2_000_000`
- The transitive order for `2 -> 3 -> 1` combines two orders, so the fee needs to be added once to the final price (to account for lost fees needed for going over an additional order).